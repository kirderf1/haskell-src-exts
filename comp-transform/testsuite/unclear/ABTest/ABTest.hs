{-# LANGUAGE ComposableTypes #-}

piececategory A
piececategory B

-- TODO What should happen if a piece has a parameter for a different category?
data piece A ==> A = A A B
data piece A ==> Ab = Ab Bool B

data piece B ==> B = B Bool | Bi Int

type CombinedA = A ==> (A, Ab)

main = putStrLn $ "Det funkar!"
