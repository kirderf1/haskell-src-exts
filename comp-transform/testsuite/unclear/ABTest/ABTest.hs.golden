{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE FlexibleContexts #-}
import qualified Data.Comp
import qualified Data.Comp.Derive
import qualified Data.Comp.Show ()
import qualified Data.Comp.Equality ()

data A a = A a B
             deriving Functor

$(
  Data.Comp.Derive.derive
    [Data.Comp.Derive.makeTraversable, Data.Comp.Derive.makeFoldable,
     Data.Comp.Derive.makeEqF, Data.Comp.Derive.makeShowF,
     Data.Comp.Derive.smartConstructors,
     Data.Comp.Derive.smartAConstructors]
    [''A]
  )

data Ab a = Ab Bool B
              deriving Functor

$(
  Data.Comp.Derive.derive
    [Data.Comp.Derive.makeTraversable, Data.Comp.Derive.makeFoldable,
     Data.Comp.Derive.makeEqF, Data.Comp.Derive.makeShowF,
     Data.Comp.Derive.smartConstructors,
     Data.Comp.Derive.smartAConstructors]
    [''Ab]
  )

data B a = B Bool
         | Bi Int
             deriving Functor

$(
  Data.Comp.Derive.derive
    [Data.Comp.Derive.makeTraversable, Data.Comp.Derive.makeFoldable,
     Data.Comp.Derive.makeEqF, Data.Comp.Derive.makeShowF,
     Data.Comp.Derive.smartConstructors,
     Data.Comp.Derive.smartAConstructors]
    [''B]
  )

type CombinedA = Data.Comp.Term (A Data.Comp.:+: Ab)
main = putStrLn $ "Det funkar!"
